### PR TITLE
ARROW-11738: [Rust][DataFusion] Fix Concat and Trim Functions

### PR DESCRIPTION
**Pull request by seddonm1**
This PR is a child of https://github.com/apache/arrow/pull/9243

It does a few things that are hard to separate:

- fixes the behavior of `concat` and `trim` functions to be in line with the Postgres implementations
- restructures some of the code base (mainly sorting and adding tests) to facilitate easier testing and implementation of the remainder of https://github.com/apache/arrow/pull/9243

@alamb @jorgecarleitao 
please review but merging will be dependent on https://github.com/apache/arrow/pull/9507

### CODE REVIEW

**Comment by seddonm1** [comment]
@Dandandan I can add `substr` to this PR if that would help you with TCPH-q22?

**Comment by seddonm1** [comment]
@Dandandan I have added `substr` to this PR

File: rust/datafusion/src/physical_plan/functions.rs [hunk 1]
```diff
@@ -171,44 +180,51 @@ impl FromStr for BuiltinScalarFunction {
...
+            "tan" => BuiltinScalarFunction::Tan,
+            "trunc" => BuiltinScalarFunction::Trunc,
+
+            // string functions
+            "array" => BuiltinScalarFunction::Array,
+            "bit_length" => BuiltinScalarFunction::BitLength,
+            "btrim" => BuiltinScalarFunction::Btrim,
+            "char_length" => BuiltinScalarFunction::CharacterLength,
+            "character_length" => BuiltinScalarFunction::CharacterLength,
+            "concat_ws" => BuiltinScalarFunction::ConcatWithSeparator,
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/functions.rs:208]: Alphabetically, this should be after `concat`?

Comment (seddonm1) [reply rust/datafusion/src/physical_plan/functions.rs]: yes, my VSCode sorting treats `_` before `"`. fixed :+1:

**Review by Dandandan — APPROVED** [review]
LGTM!
I think from time to time we will can check on the status of the tpch queries. Feel free to split the PRs in the way you think makes sense @seddonm1 👍

**Comment by seddonm1** [comment]
> LGTM!
> I think from time to time we will can check on the status of the tpch queries. Feel free to split the PRs in the way you think makes sense @seddonm1 👍

@Dandandan basically the string functions PR is so big even splitting like this makes a big difference and allows applying some of the foundational components (see the testing) required for the rest. So if this helps unblock you then thats a good thing :D

**Review by alamb — APPROVED** [review]
Epic work @seddonm1  -- thank you again. I skimmed this and spot checked a few functions, but it all looks as good as https://github.com/apache/arrow/pull/9243 is. I think this one is good to merge.

**Comment by alamb** [comment]
This PR sadly does need a rebase

**Comment by seddonm1** [comment]
thanks @alamb rebased 👍

**Comment by alamb** [comment]
Integration test failure looks like https://issues.apache.org/jira/browse/ARROW-11717, so ignoring

**Pull request closed without merging**